### PR TITLE
Add spotlight capability to tour

### LINKED ISSUE #3075: Add spotlight/pointing capability to tours
Consider: https://github.com/gilbarbara/react-joyride and https://github.com/gilbarbara/react-joyride/blob/master/src/components/Spotlight.js

### PR DESCRIPTION
**Pull request by jasontk19**
## Description

Fixes #3075.

* Integrate React Joyride
* Allow configuration via tour steps
* Add ability to flag tours as "mock" so they don't show in the tour list to users
* Add mock tour to run tests against and serve as example config

## How to test

1. Load the app via this URL to load the mock tour: http://localhost:3000/?tr=hurricane_dorian_september_joyride&mockTour=true
2. Advance to a tour step with a Joyride beacon showing
3. Click the beacon to open the associated tooltip
4. Confirm closing the tooltip moves to the next step
5. Click the next beacon to open the next tooltip
6. Click through next/prev to confirm navigation between steps
7. Advance to the last step and click "last"
8. Confirm tooltip closes and no beacons showing
9. Advance to mock tour step 4
10. Resize browser window and confirm beacon location on map updates 
11. Click beacon to open tooltip
12. Confirm hovering mouse over the spotlight area allows mouse interactions to be picked up by the map
13. Resize browser window and confirm spotlight area on map updates location

### CODE REVIEW

File: web/js/app.js [hunk 1]
```diff
@@ -191,7 +192,7 @@ class App extends React.Component {
         <div id="layer-settings-modal" />
         <div id="eventsHolder" />
         <div id="imagedownload" />
-        <Timeline key={locationKey || '1'} />
+        <Timeline />
```
[the remark this answers is not in the document]
Comment (Benjaki2) [reply web/js/app.js:195]: Looks okay to me 🤷

**Review by Benjaki2 — COMMENTED** [review]
Really like the look of the spotlight! 💡 

I was able to get it to offset a bit by resizing the the browser (making it smaller):
Screen Shot 2020-10-07 at 11 30 39 AM

Besides that, it looks great to me!

**Review by edplato — APPROVED** [review]
Looks great, works well 🎯 . I have a couple minor change requests:

- Switching to a polar projection causes the beacon to position in undesired spots. You could probably recalculate/transform coordinates to the changed projection, but for this feature use case I would recommend preventing the beacon from carrying over to different projections and limit it to it's original projection that the tour step was on.
- I prefer no box-shadow added in the override css - half styling/half performance related

File: web/js/app.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (edplato) [reply web/js/app.js:195]: 🤷‍♂️ I remember that was needed at the time, but looks fine here now

**Comment by jasontk19** [comment]
Pushed updates that (should) handle the element re-positioning more reliably on browser resize or map move/zoom events.
  
Also updated so that the Joyride is removed entirely if the user switching to a projection other than the one that is defined in the tour step's `stepLink`.  It is reset if they switch back to the step's projection.

@edplato can you clarify on the `box-shadow` performance concern?  It looks like animating that property can be a repaint issue but it looks like the Joyride CSS is only animating the opacity property which, [if I understand correctly is the way to go](https://tobiasahlin.com/blog/how-to-animate-box-shadow/)?  

As for the looks of the shadow, I could go either way.  I was worried people would think the beacon doesn't stand out enough in certain situations. (it can get a little lost on the map but maybe it's not really an issue)

**Comment by edplato** [comment]
@jasontk19 I think you're definitely handling the animation right. I was originally noticing choppiness on the scale transform that went away when I removed the `box-shadow` from the animation, but now I'm not seeing any issues and it's very smooth with or without the `box-shadow`. It must've been an issue on my end with too many browsers open while testing. Not quite sure.

The projection switch change looks/works well. I like that is resets when changing back to the `stepLink` projection.

***
I'm running into a map repositioning/hiding the beacon issue. Below are some steps to reproduce.


**Steps:**
1. Go to step 4/6 of the test tour
2. Drag the map down (so north is more in view) then the beacon is removed from the map view after the target storm is no longer within the map view
3. Drag the map up (so south is more in view) past the target storm area then the beacon remains positioned on the top of the map view. There may need to be some tweaking to make sure the beacon is removed when the target coordinates are out of view.


***
I found something new that may be related to map repositioning/Joyride state.  Below are some steps to reproduce.

**Steps:**
1. Go to step 4/6 of the test tour
2. Click the first beacon (located on the map)
3. Click next button and the second beacon (located in the sidebar) shows up
4. Reposition the map and the first beacon (located on the map) shows back up and the Joyride steps seem to restart. I would expect the Joyride step to be retained in this case.

***
Extra thought, what do you think about preventing closing the step by clicking behind the Joyride steps, so the only way to get rid of them is to click `x` or work through the steps? Setting `disableOverlayClose` to `true` should handle this. I think it may be more consistent in respect to how the current tour steps work where you can click around the map/interface.

**Comment by jasontk19** [comment]
> I'm running into a map repositioning/hiding the beacon issue. Below are some steps to reproduce.
> 
> **Steps:**
> 
> 1. Go to step 4/6 of the test tour
> 2. Drag the map down (so north is more in view) then the beacon is removed from the map view after the target storm is no longer within the map view
> 3. Drag the map up (so south is more in view) past the target storm area then the beacon remains positioned on the top of the map view. There may need to be some tweaking to make sure the beacon is removed when the target coordinates are out of view.
> 

I don't think there's much I can do about this without re-writing our own beacon component from scratch.  The beacon gets placed by the Joyride library wherever the target element is and the API doesn't provide much in the way of customizing beacon behavior or positioning like this.  We could consider using the "fly" animation of the map to always move it to center over the element but honestly, I'd rather wait and see if/how much this functionality gets used before putting too much time into this. 

> I found something new that may be related to map repositioning/Joyride state. Below are some steps to reproduce.
> 
> **Steps:**
> 
> 1. Go to step 4/6 of the test tour
> 2. Click the first beacon (located on the map)
> 3. Click next button and the second beacon (located in the sidebar) shows up
> 4. Reposition the map and the first beacon (located on the map) shows back up and the Joyride steps seem to restart. I would expect the Joyride step to be retained in this case.

Ok, it took a bit of fiddling, but I think this should be fixed now.  Unfortunately, because of some expectations/limitations inherent in the Joyride `<Step>` component code, I had to disable the `continuous` option for any Joyride tours that had map targets beyond step one in order for this to work right.  Stepping forward/backward between Joyride steps or tour steps now, I'm not seeing a reliable way to get the positioning of the beacon to be incorrect.  Let me know how it holds up for you!

> Extra thought, what do you think about preventing closing the step by clicking behind the Joyride steps, so the only way to get rid of them is to click `x` or work through the steps? Setting `disableOverlayClose` to `true` should handle this. I think it may be more consistent in respect to how the current tour steps work where you can click around the map/interface.

[I am including this as an optional tour config parameter](https://github.com/nasa-gibs/worldview/pull/3155/files#diff-e1776cd01654365cae8239cbcf0e0fe61d1840139e6c0b8e6d3c4cbd2f556f59R191) since I wasn't sure how tour authors, e.g. @minniewong, would want it to work. I probably need to write some documentation about what params are supported and how!  Do we have any existing docs around writing tours?

**Comment by edplato** [comment]
@jasontk19 Looks good to me. I see your parameters to handle `disableOverlayClose`, so more options the better for flexibility in making tours. Here is the tour story documentation that I know of: [https://github.com/nasa-gibs/worldview/blob/master/doc/configuration.md#new-tour-stories](https://github.com/nasa-gibs/worldview/blob/master/doc/configuration.md#new-tour-stories).

**Review by Benjaki2 — APPROVED** [review]
Looks good to me. 👍  I was still able to get the spotlight stuck in the wrong place but only if I tried really hard.

**Review by edplato — APPROVED** [review]
👍

**Pull request merged**